src/builder.ts

```diff
@@ -95,56 +106,58 @@ export default class Builder {
 							if (data == null) {
 								return;
 							}
-							if (data.error) {
-								errored = true;
-								dup.destroy(new Error(data.error));
-							} else {
-								// Store image layers, so that they can be deleted by the caller
-								// if necessary
-								const sha = Utils.extractLayer(data.stream);
-								if (sha !== undefined) {
-									self.layers.push(sha);
+							try {
+								if (data.error) {
+									throw new Error(data.error);
+								} else {
+									// Store image layers, so that they can be
+									// deleted by the caller if necessary
+									const sha = Utils.extractLayer(data.stream);
+									if (sha !== undefined) {
+										self.layers.push(sha);
+									}
+									this.emit('data', data.stream);
 								}
-
-								this.emit('data', data.stream);
+							} catch (error) {
+								daemonStream.unpipe();
+								reject(error);
```

Comment:
* `unpipe` is called to stop the unnecessary further processing of the docker daemon's output once an error is caught and the build rejected.
* `reject(error)` will eventually cause `failBuild` to be called.